code/__HELPERS/path.dm

```diff
@@ -0,0 +1,328 @@
+#define PATH_DIST(A, B) (A.Distance(B))
+/// Enumerator for the starting turf's sources value, so we know when we've hit the beginning when unwinding at the end
+#define PATH_START	-1
+
+/**
+ * This file contains the stuff you need for using JPS (Jump Point Search) pathing, an alternative to A* that skips
+ * over large numbers of uninteresting tiles resulting in much quicker pathfinding solutions.
+ *
+ * Quick notes about current implementation:
+ * * JPS requires and allows for diagonal movement, whereas our A* only allows for cardinal movement
+ * *
+ * *
+ * *
+ */
+
+
+//JPS nodes variables
+/datum/jpsnode
+	/// The turf associated with this node
+	var/turf/tile
+	/// The node we just came from
+	var/datum/jpsnode/prevNode
+	/// The A* node weight (f = number_of_tiles + heuristic)
+	var/f
+	/// The A* node heuristic (a rough estimate of how far we are from the goal)
+	var/h
+	/// How many steps it's taken to get here from the start (currently pulling double duty as steps taken & cost to get here, since all moves incl diagonals cost 1 rn)
+	var/nt
+	/// How many steps it took to get here from the last node
+	var/jumps
+	/// Nodes store the endgoal so they can process their heuristic without a reference to the pathfind datum
+	var/turf/node_goal
+
+/datum/jpsnode/New(turf/our_tile, datum/jpsnode/previous_node, jumps_taken, turf/incoming_goal)
+	tile = our_tile
+	prevNode = previous_node
+	jumps = jumps_taken
+	if(prevNode)
+		nt = prevNode.nt + jumps
+		node_goal = prevNode.node_goal
+	else
+		nt = 0
+		node_goal = incoming_goal
+
+	h = PATH_DIST(tile, node_goal)
+	f = nt + h//*(1+ PF_TIEBREAKER)
+
+/datum/pathfind
+	/// The thing that we're actually trying to path for
+	var/atom/movable/caller
+	/// The turf where we started at
+	var/turf/start
+	/// The turf we're trying to path to (note that this won't track a moving target)
+	var/turf/end
+	/// The open list/stack we pop nodes out from
+	var/datum/heap/open
+	/// An assoc list that matches turfs (the key) to their nodes if said turf has one
+	var/list/openc
+	/**
+	 * An assoc list that serves as the closed list & tracks what turfs came from where. Key is the turf, and the value is what turf it came from
+	 *
+	 * Nodes are only created & added to the heap/openc once a turf is found "interesting", but due to recursion, we may not know it's interesting until we finish processing its children.
+	 * Inserting to this list is cheaper than making a node datum + inserting into the heap, so everything goes in here immediately, and what we use to recreate the path at the end
+	 */
+	var/list/sources
+	/// The list we compile at the end if successful to pass back
+	var/list/path
+
+	// general pathfinding vars/args
+	/// An ID card representing what access we have and what doors we can open. Its location relative to the pathing atom is irrelevant
+	var/obj/item/card/id/id
```

Comment:
this is not really ideal
is this how the previous system worked?

Reply:
on goon atleast, yes